File: Python/sortedSquares.py
from typing import List


# 给你一个按 非递减顺序 排序的整数数组 nums，返回 每个数字的平方 组成的新数组，要求也按 非递减顺序 排序

# def sortedSquares(nums):
#     for i in range(len(nums)):
#         nums[i] *= nums[i]
#     return nums.sort()

def sortedSquares(nums: list[int]) -> list[int]:
    i = 0
    j, k = len(nums) - 1, len(nums) - 1
    result = [float('inf')] * len(nums) # float('inf')代表正无穷
    while i <= j:
        if nums[i] * nums[i] > nums[j] * nums[j]:
            result[k] = nums[i] * nums[i]
            i += 1
        else:
            result[k] = nums[j] * nums[j]
            j -= 1
        k -= 1
    return result


# nums = [-4, -1, 0, 3, 10]
nums = [-7, -3, 2, 3, 11]
res = sortedSquares(nums)
print(res)

File: Java_ALG/src/Array/SortedSquares.java
package Array;

import java.util.Arrays;

/**
 * @program: alghorithm
 * @description: 有序数组的平方
 * @author: wangzijin
 * @create: 2024-02-19 20:17
 **/
// https://leetcode.cn/problems/squares-of-a-sorted-array/description/
class SortedSquares {
    public int[] method(int[] nums) {
        /**
         * 用双指针，不要使用Arrays.sort(nums)
         * 暴力排序时间复杂度：O(n + nlogN)
         * 双指针时间复杂度：O(n)
         */
        int length = nums.length;
        int[] result = new int[length];
        int left = 0, right = length - 1;
        int index = length - 1;
        while (left <= right) {
            if (nums[left] * nums[left] > nums[right] * nums[right]) {
                result[index--] = nums[left] * nums[left];
                left++;
            } else {
                result[index--] = nums[right] * nums[right];
                right--;
            }
        }
        return result;
    }

    public static void main(String[] args) {
        int[] nums = {-4,-1,0,3,10};
        SortedSquares test = new SortedSquares();
        int[] res = test.method(nums);
        System.out.println(Arrays.toString(res));
    }
}